apps/www/content/docs/dark-mode/tanstack.mdx

```diff
@@ -0,0 +1,147 @@
+---
+title: Tanstack Start
+description: Adding dark mode to your TanStack Start app.
+---
+
+## Dark mode
+
+<Steps>
+
+### Create server functions
+
+Start by creating `serverFn`s to get and set the theme in cookies.
+
+```ts title="lib/theme.ts"
+import { type Theme } from "@/components/theme-provider";
+import { createServerFn } from "@tanstack/react-start";
+import { getCookie, setCookie } from "@tanstack/react-start/server";
+
+const storageKey = "ui-theme";
+
+export const getThemeServerFn = createServerFn().handler(async () => {
+  return (getCookie(storageKey) || "light") as Theme;
+});
+
+export const setThemeServerFn = createServerFn({ method: "POST" })
+  .validator((data: unknown) => {
+    if (typeof data != "string" || (data != "dark" && data != "light")) {
+      throw new Error("Invalid theme provided");
+    }
+    return data as Theme;
+  })
+  .handler(async ({ data }) => {
+    setCookie(storageKey, data);
+  });
+```
+
+### Create a theme provider
+
+```tsx title="components/theme-provider.tsx"
+import { setThemeServerFn } from "@/lib/theme";
+import { useRouter } from "@tanstack/react-router";
+import { createContext, PropsWithChildren, use } from "react";
+
+export type Theme = "light" | "dark";
+
+type ThemeContextVal = { theme: Theme; setTheme: (val: Theme) => void };
+type Props = PropsWithChildren<{ theme: Theme }>;
+
+const ThemeContext = createContext<ThemeContextVal | null>(null);
+
+export function ThemeProvider({ children, theme }: Props) {
+  const router = useRouter();
+
+  function setTheme(val: Theme) {
+    setThemeServerFn({ data: val });
+    router.invalidate();
+  }
+
+  return <ThemeContext value={{ theme, setTheme }}>{children}</ThemeContext>;
+}
+
+export function useTheme() {
+  const val = use(ThemeContext);
+  if (!val) throw new Error("useTheme called outside of ThemeProvider!");
+  return val;
+}
+```
+
+### Wrap your root layout
+
+Wrap your app in the `ThemeProvider` and pass the theme to the `html` tag using `className`.
+
+```tsx {1,10,12,21,23} title="routes/__root.tsx"
```

Comment:
It could be a lot more useful to show a diff view here instead as I missed a single line to change when comparing this to my file. An example of it being used in the docs already is here: https://ui.shadcn.com/docs/components/pagination#nextjs
https://github.com/shadcn-ui/ui/blob/04432835f91dbe0a3273f50ac829e0a377110e5f/apps/v4/content/docs/components/pagination.mdx?plain=1#L83-L99